tests/unit/test_charm.py

```diff
@@ -316,163 +307,104 @@ def test_unit_ips(self):
 
     @patch_network_get(private_address="1.1.1.1")
+    @patch("ops.framework.EventBase.defer")
     @patch("charm.MongoDBConnection")
-    @patch("charm.MongodbOperatorCharm._need_replica_set_reconfiguration", return_value=True)
-    @patch("mongod_helpers.MongoDB.reconfigure_replica_set")
-    @patch("mongod_helpers.MongoDB.check_replica_status")
-    @patch("charm.MongodbOperatorCharm._replica_set_hosts")
-    def test_mongodb_relation_joined_check_status_failure(
-        self, _, check_replica_status, mongodb_reconfigure, need_reconfiguration, connection
-    ):
-        """Test failure in checking status results in waiting.
+    @patch("lib.charms.mongodb_libs.v0.mongodb.MongoClient")
+    def test_relation_joined_get_members_failure(self, client, connection, defer):
+        """Tests reconfigure does not execute when unable to get the replica set members.
 
-        Tests the scenario that a failure to check the current statuses of replica set hosts
-        results in WaitingStatus and no attempt to reconfigure is made.
+        Verifies in case of relation_joined and relation departed, that when the the database
+        cannot retrieve the replica set members that no attempts to remove/add units are made and
+        that the the event is deferred.
         """
-        # preset values
-        self.harness.charm._replica_set_hosts = ["1.1.1.1"]
+        # presets
         self.harness.set_leader(True)
-        connection.return_value.__enter__.return_value.is_ready = True
+        rel = self.harness.charm.model.get_relation("mongodb")
 
         for exception in PYMONGO_EXCEPTIONS:
-            # simulate failure to check replica status
-            check_replica_status.side_effect = exception
+            connection.return_value.__enter__.return_value.get_replset_members.side_effect = (
+                exception
+            )
 
-            # simulate 2nd MongoDB unit
-            rel = self.harness.charm.model.get_relation("mongodb")
-            self.harness.add_relation_unit(rel.id, "mongodb/1")
-            self.harness.update_relation_data(rel.id, "mongodb/1", PEER_ADDR)
+            # test both relation events
+            for departed in [False, True]:
+                if departed:
+                    # simulate removing 2nd MongoDB unit
+                    self.harness.remove_relation_unit(rel.id, "mongodb/1")
+                else:
+                    # simulate 2nd MongoDB unit joining
+                    self.harness.add_relation_unit(rel.id, "mongodb/1")
+                    self.harness.update_relation_data(rel.id, "mongodb/1", PEER_ADDR)
 
-            # verify we go into waiting and don't reconfigure
-            self.assertEqual(
-                self.harness.charm.unit.status, WaitingStatus("waiting to reconfigure replica set")
-            )
-            mongodb_reconfigure.assert_not_called()
+                if departed:
```

Comment:
One question: why do you need to have this if/else block separated from the one in L364? (the condition is the same).

Reply:
You're so smart!

They shouldn't be separated :)